src/compiler/types.ts

```diff
@@ -2089,6 +2108,7 @@ namespace ts {
         ESSymbol                = 0x01000000,  // Type of symbol primitive introduced in ES6
         ThisType                = 0x02000000,  // This type
         ObjectLiteralPatternWithComputedProperties = 0x04000000,  // Object literal type implied by binding pattern has computed properties
+        PredicateType           = 0x08000000,

```

Comment:
Why special case type predicates everywhere with extra slots for them when they're mutually exclusive with a return type (if there is a type predicate, the return type _must_ be boolean)? They even affect type equality (differently than booleans do) when they're nested within a type. Why not unify them _all_ as predicate types? It would be fairly easy to unify them in this way, IMO.

Yes, it being a type is why it follows assignments at present (like all types), but it shouldn't be hard to fix - it's just a matter of making the right branches of `getWidenedType` return `boolean` for type predicates.

The more interesting issue I've noted with them right now is that `this` types aren't properly reinstantiated when narrowing. I can't nest two `this` type guards, like so:
```ts
class FSO {
  isFile: this is File;
  isNetworked: this is (this & Networked);
}
class File {
  path: string;
}
interface Networked {
  host: string;
}
let f: FSO = new File();
if (f.isFile) {
  // f should be File
  if (f.isNetworked) {
    // f should be File & Networked, but is just File 
  }
}
```